src/services/executionEnvironment.ts

```diff
@@ -0,0 +1,308 @@
+import * as child_process from 'child_process';
+import * as fs from 'fs';
+import * as path from 'path';
+import { URI } from 'vscode-uri';
+import { Connection } from 'vscode-languageserver';
+import { v4 as uuidv4 } from 'uuid';
+import { ImagePuller } from '../utils/imagePuller';
+import { asyncExec } from '../utils/misc';
+import { WorkspaceFolderContext } from './workspaceManager';
+import { IContainerEngine } from '../interfaces/extensionSettings';
+
+export class ExecutionEnvironment {
+  private connection: Connection;
+  private context: WorkspaceFolderContext;
+  private _container_engine: IContainerEngine;
+  private _container_image: string;
+  private _container_image_id: string;
+
+  constructor(connection: Connection, context: WorkspaceFolderContext) {
+    this.connection = connection;
+    this.context = context;
+  }
+
+  public async initialize(): Promise<void> {
+    try {
+      const settings = await this.context.documentSettings.get(
+        this.context.workspaceFolder.uri
+      );
+      this._container_image = settings.executionEnvironment.image;
+      this._container_engine = settings.executionEnvironment.containerEngine;
+      if (this._container_engine === 'auto') {
+        for (const ce of ['podman', 'docker']) {
+          try {
+            child_process.execSync(`which ${ce}`, {
+              encoding: 'utf-8',
+            });
+            this._container_engine = <IContainerEngine>ce;
+            this.connection.console.log(`Container engine set to: '${ce}'`);
+            break;
+          } catch (error) {
+            this.connection.console.info(`Container engine '${ce}' not found`);
+          }
```

Comment:
This way it's clear that only the subprocess invocation is expected to error out 
```suggestion
          try {
            child_process.execSync(`which ${ce}`, {
              encoding: 'utf-8',
            });
          } catch (error) {
            this.connection.console.info(`Container engine '${ce}' not found`);
            continue;
          }
          this._container_engine = <IContainerEngine>ce;
          this.connection.console.log(`Container engine set to: '${ce}'`);
          break;
```

Reply:
I still feel the exiting code in PR is more readable :-), but I do appreciate your review and if you think your approach is more cleaner I will fix the review comment.

Reply:
Fixed